### PR TITLE
Set PCTASKS_COSMOSDB__TEST_CONTAINER_SUFFIX in Functions

### PR DESCRIPTION
**Pull request by TomAugspurger**
This sets the `PCTASKS_COSMOSDB__TEST_CONTAINER_SUFFIX` environment
variable in the Azure Functions, to ensure that the test Cosmos DB
containers are used.

We're using Azure Function's "Binding expression patterns"
(https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-expressions-patterns)
to dynamically set the Cosmos DB container to watch. This required
updating the terraform to set the container name to watch as an app
setting.

This did require a small change to the names of the temporary Cosmos DB
containers we create for testing. Previously they were
`tmp-{name}-{suffix}-{uuid}`.

To make templating possible in the docker-compose file, I went with
`{name}{suffix}`, where suffix is uniquely set to the GitHub actions run
ID, so should be sufficient for avoiding clashes.

In https://github.com/microsoft/planetary-computer-tasks/actions/runs/5023839052/jobs/9008884389, this failed. I'm not sure why,
but locally I could reproduce it with

```
pctasks-server               | Message: Message: {"Errors":["Resource Not Found. Learn more: https:\/\/aka.ms\/cosmosdb-tsg-not-found"]}
pctasks-server               | ActivityId: 66e88ca9-8a06-4647-87f7-6c7e51d509b0, Request URI: /apps/8e279c36-21d7-4ec2-ba9e-b46adbf5ccba/services/709051fd-af4c-461b-83ae-a3c3fe625901/partitions/e2862f23-70eb-453a-910c-32298d3a9eba/replicas/133082039478945105s, RequestStats:
```

This makes me think that `pctasks-server` has a different view of what the temporary test container is called.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
I'm not sure why CI is failing here. The error message doesn't give much information. Just a status code.

```
        if http_error_msg:
>           raise HTTPError(http_error_msg, response=self)
E           requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://nginx/tasks/workflows/unittest-workflow-test_client_submit
```

I don't think we have access to the server containers stdout.

I can reproduce this locally by deleting my test containers

```
python pctasks/dev/pctasks/dev/cosmosdb.py rm-test-containers
Deleting container: workflow-runstest-tom
Deleting container: workflowstest-tom
Deleting container: recordstest-tom
```

And running the test

```
» pytest pctasks/client/tests/test_client.py::test_client_submit --tb=line
======================================================================================================================================= test session starts =======================================================================================================================================
...
pctasks/client/tests/test_client.py F                                                                                                                                                                                                                                                       [100%]

============================================================================================================================================ FAILURES =============================================================================================================================================
/home/taugspurger/src/Microsoft/planetary-computer-tasks/.direnv/python-3.10.8/lib/python3.10/site-packages/requests/models.py:1021: requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://localhost:8510/tasks/workflows/unittest-workflow-test_client_submit
...
```

Locally, I see that the pctasks-server container is trying to write the document and getting a ResourceNotFoundError

```
pctasks-server               |   File "/opt/src/server/pctasks/server/routes/workflows.py", line 102, in create_workflow
pctasks-server               |     await container.put(record)
pctasks-server               |   File "/opt/src/core/pctasks/core/cosmos/container.py", line 456, in put
pctasks-server               |     await with_backoff_async(
pctasks-server               |   File "/opt/src/core/pctasks/core/utils/backoff.py", line 177, in with_backoff_async
pctasks-server               |     return await fn()
pctasks-server               |   File "/opt/src/core/pctasks/core/cosmos/container.py", line 446, in _upsert
pctasks-server               |     return await self._container_client.upsert_item(
...
pctasks-server               |     raise exceptions.CosmosResourceNotFoundError(message=data, response=response)
pctasks-server               | azure.cosmos.exceptions.CosmosResourceNotFoundError: (NotFound) Message: {"Errors":["Resource Not Found. Learn more: https:\/\/aka.ms\/cosmosdb-tsg-not-found"]}
```

I need to determine: what container is pctasks-server trying to write to, and why it thinks it doesn't exist.

**Comment by TomAugspurger** [comment]
805bae2 added the output of `docker-compose logs`. This shows that the Azure Functions is failing to start because we try to start it before the Cosmos DB database is ready:

```
pctasks-functions        | fail: Host.Startup[0]
pctasks-functions        |       The listener for function 'Functions.WorkflowsCF' was unable to start.
pctasks-functions        |       Microsoft.Azure.WebJobs.Host.Listeners.FunctionListenerException: The listener for function 'Functions.WorkflowsCF' was unable to start.
pctasks-functions        |        ---> System.InvalidOperationException: Either the source collection 'workflows5027005382' (in database 'pctasks')  or the lease collection 'leases' (in database 'pctasks') does not exist. Both collections must exist before the listener starts. To automatically create the lease collection, set 'CreateLeaseCollectionIfNotExists' to 'true'.
```

That said, I am a *bit* surprised by this error message from pctasks-server. 

```
pctasks-server           | INFO:     172.16.238.9:44358 - "POST /tasks/code/upload HTTP/1.0" 200 OK
pctasks-server           | [INFO] 2023-05-19 18:26:46,159 - Create workflow: unittest-workflow-test_client_submit {'ref_id': None, 'request_entity': None, 'service': 'tasks', 'http.url': 'http://nginx/tasks/workflows/unittest-workflow-test_client_submit', 'http.method': 'POST', 'http.path': '/tasks/workflows/unittest-workflow-test_client_submit', 'request_id': None, 'subscription_key': 'hunter2', 'user_email': 'dev@null.com', 'request_url': 'http://nginx/tasks/workflows/unittest-workflow-test_client_submit', 'workflow': 'unittest-workflow-test_client_submit'}
pctasks-server           | [ERROR] 2023-05-19 18:26:46,199 - Exception when handling request {'ref_id': None, 'request_entity': None, 'service': 'tasks', 'http.url': 'http://nginx/tasks/workflows/unittest-workflow-test_client_submit', 'http.method': 'POST', 'http.path': '/tasks/workflows/unittest-workflow-test_client_submit', 'request_id': None, 'subscription_key': 'hunter2', 'user_email': 'dev@null.com', 'request_url': 'http://nginx/tasks/workflows/unittest-workflow-test_client_submit', 'stackTrace': '(NotFound) Message: {"Errors":["Resource Not Found. Learn more: https:\\/\\/aka.ms\\/cosmosdb-tsg-not-found"]}\r\nActivityId: 0d57ca40-146d-40ba-a334-d9ec36f77ec0, Request URI: /apps/8e279c36-21d7-4ec2-ba9e-b46adbf5ccba/services/709051fd-af4c-461b-83ae-a3c3fe625901/partitions/e2862f23-70eb-453a-910c-32298d3a9eba/replicas/133082039478945105s, 
```

I would have expected all the containers to exist by this point (see [here](https://github.com/microsoft/planetary-computer-tasks/actions/runs/5027005382/jobs/9015968106?pr=207#step:5:12463)).


```
Creating planetary-computer-tasks_dev_run ... done
Creating database...
Creating containers...
 - workflow-runs5027005382
 - workflows5027005382
 - records5027005382
Creating lease container...
Creating stored procedures...
  - Container: workflow-runs5027005382
     bulkput-workflowruns (create)
Creating triggers...
  - Container: workflows5027005382
     post-all-workflows (create)
  - Container: workflow-runs5027005382
     post-all-workflowruns (create)
```

But perhaps I'm misreading the situation.

**Comment by TomAugspurger** [comment]
0cff483901264926edfec4a62d75ea868d2e8a83 is another attempt at starting the Azure Functions after doing the Cosmos DB setup. https://github.com/microsoft/planetary-computer-tasks/pull/207#issuecomment-1554762088 was an earlier attempt but that failed because the Cosmos setup runs along with some code that requires Azurite to be running. So we can't do *all* of the server setup after the Cosmos DB setup.

A potential option, tried here, is to split the Azure Functions stuff to a separate docker-compose file. In `scripts/setup` we'll

1. Start all the servers except the Functions (`--no-azure-functions`)
2. Do the Cosmos setup (which should succeed, since azurite is running)
3. Start the Azure Functions servers

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -53,6 +53,10 @@ jobs:
         run: ./scripts/citest-integration
         if: ${{ github.base_ref  == 'main' }}
 
+      - name: Get docker-compose logs
```
Comment (TomAugspurger) [inline .github/workflows/pr.yml:56]: Dunno if we want this here always, but it's somewhat handy for debugging.

**Comment by TomAugspurger** [comment]
Bummer. CI failed again, with the same error

```
pctasks-server           | [INFO] 2023-05-19 20:02:44,552 - Create workflow: unittest-workflow-test_client_submit {'ref_id': None, 'request_entity': None, 'service': 'tasks', 'http.url': 'http://nginx/tasks/workflows/unittest-workflow-test_client_submit', 'http.method': 'POST', 'http.path': '/tasks/workflows/unittest-workflow-test_client_submit', 'request_id': None, 'subscription_key': 'hunter2', 'user_email': 'dev@null.com', 'request_url': 'http://nginx/tasks/workflows/unittest-workflow-test_client_submit', 'workflow': 'unittest-workflow-test_client_submit'}
pctasks-server           | [ERROR] 2023-05-19 20:02:44,798 - Exception when handling request {'ref_id': None, 'request_entity': None, 'service': 'tasks', 'http.url': 'http://nginx/tasks/workflows/unittest-workflow-test_client_submit', 'http.method': 'POST', 'http.path': '/tasks/workflows/unittest-workflow-test_client_submit', 'request_id': None, 'subscription_key': 'hunter2', 'user_email': 'dev@null.com', 'request_url': 'http://nginx/tasks/workflows/unittest-workflow-test_client_submit', 'stackTrace': '(NotFound) Message: {"Errors":["Resource Not Found. Learn more: https:\\/\\/aka.ms\\/cosmosdb-tsg-not-found"]}
```

I'll try reorganizing things one more time, to split the Cosmos DB setup out of the rest of the dev setup, rather than splitting Azure Functions out of the servers :/

**Comment by TomAugspurger** [comment]
Hmm I really thought https://github.com/microsoft/planetary-computer-tasks/pull/207/commits/ecc399b7ce2b52d686b821c0fabaef4fb98c137e was going to pass, but it failed with the same error as earlier. A 500 error on the test, and the pctasks-server indicating that it couldn't write to the storage container.

File: pctasks/server/pctasks/server/routes/workflows.py [hunk 2]
```diff
@@ -99,6 +99,8 @@ async def create_workflow(
         if existing:
             raise HTTPException(status_code=409, detail="Workflow already exists")
 
+        logger.warning("Writing to %s", container.name)
+
```
Comment (TomAugspurger) [inline pctasks/server/pctasks/server/routes/workflows.py:103]:
```suggestion
```

File: pctasks/core/pctasks/core/cosmos/settings.py [hunk 3]
```diff
@@ -46,35 +46,23 @@ def section_name(cls) -> str:
...
 
     def get_records_container_name(self) -> str:
-        if self.test_container_suffix:
-            return f"tmp-{self.records_container_name}-{self.test_container_suffix}"
-        return self.records_container_name
+        return f"{self.records_container_name}{self.test_container_suffix}"
+
+    @validator("test_container_suffix", always=True)
+    def _validate_test_container_suffix(cls, v: Optional[str]) -> str:
+        return v or ""
```
Comment (lossyrob) [inline pctasks/core/pctasks/core/cosmos/settings.py:65]: I think this is fine here, but I've been bitten before by validators changing the values of properties and that being super confusing during debugging/tracing logic. In that case it was a more substantial change (I think changed the type from the stated type), but in this case you're just ensuring its a string value for internal logic of the methods. Might be worth a comment to say that a validated model will never actually have a None value.

**Review by lossyrob — APPROVED** [review]

**Pull request merged**